compiler/literal/src/escape.rs

```diff
@@ -35,20 +37,32 @@ pub struct EscapeLayout {
     pub len: Option<usize>,
 }
 
-pub trait Escape {
+/// Represents string types that can be escape-printed.
+///
+/// # Safety
+///
+/// `source_len` and `layout` must be accurate, and `layout.len` must not be equal
+/// to `Some(source_len)` if the string contains non-printable characters.
+pub unsafe trait Escape {
     fn source_len(&self) -> usize;
     fn layout(&self) -> &EscapeLayout;
     fn changed(&self) -> bool {
         self.layout().len != Some(self.source_len())
     }
 
-    fn write_source(&self, formatter: &mut impl std::fmt::Write) -> std::fmt::Result;
+    /// Write the body of the string directly to the formatter.
+    ///
+    /// # Safety
+    ///
+    /// This string must only contain printable characters.
+    unsafe fn write_source(&self, formatter: &mut impl std::fmt::Write) -> std::fmt::Result;
```

Comment:
I am confused about this change.
I expect `UnicodeEscape::new_repr(self.as_wtf8())` always safely print the repr. Then the surrogate must be checked somewhere before printing. Doesn't it mean we always print &str, but not wtf8?

Reply:
We always print `str`, yes. I did this because we were already treating `write_source` as an unsafe function, per the preëxisting comment in `AsciiEscape::write_source`:

```rust
            // SAFETY: this function must be called only when source is printable ascii characters
```

That was technically unsound, since anybody could just call `AsciiEscape::new_repr(b"\xff").write_source()` and invoke library UB.